ocaml/xapi-storage-script/main.ml

```diff
@@ -11,18 +11,293 @@
+
+let ( >>| ) = Fun.flip Lwt.map
+
+let ( >>>= ) = Lwt_result.bind
+
+let return = Lwt_result.return
+
+let fail = Lwt_result.fail
+
+let ( // ) = Filename.concat
+
+module Deferred = struct
+  let all_unit (lst : unit Lwt.t list) : unit Lwt.t =
+    let (_ : unit list Lwt.t) = Lwt.all lst in
+    Lwt.return ()
+
+  let errorf fmt =
+    Printf.ksprintf (fun m -> Lwt.return (Base.Or_error.error_string m)) fmt
+
+  let combine_errors lst = Lwt.all lst >>| Base.Or_error.combine_errors
+
+  let try_with f = Lwt.try_bind f return fail
+end
+
+module Sys = struct
+  type file = Regular | Directory | Other | Missing | Unknown
+
+  let file_kind ~follow_symlinks path =
+    Lwt.try_bind
+      (fun () ->
+        ( if follow_symlinks then
+            Lwt_unix.LargeFile.stat
+          else
+            Lwt_unix.LargeFile.lstat
+        )
+          path
+      )
+      (function
+        | s -> (
+          match s.Unix.LargeFile.st_kind with
+          | Unix.S_REG ->
+              Lwt.return Regular
+          | Unix.S_DIR ->
+              Lwt.return Directory
+          | _ ->
+              Lwt.return Other
+        )
+        )
+      (function
+        | Unix.Unix_error (Unix.ENOENT, _, _) ->
+            Lwt.return Missing
+        | Unix.Unix_error ((Unix.EACCES | Unix.ELOOP), _, _) ->
+            Lwt.return Unknown
+        | e ->
+            Lwt.fail e
+        )
+
+  let access path modes =
+    Lwt.try_bind
+      (fun () -> Lwt_unix.access path modes)
+      return
+      (fun exn -> fail (`not_executable (path, exn)))
+
+  let assert_is_executable path =
+    file_kind ~follow_symlinks:true path >>= function
+    | Directory | Other | Missing | Unknown ->
+        fail (`missing path)
+    | Regular -> (
+        access path [Unix.X_OK] >>= function
+        | Error exn ->
+            fail exn
+        | Ok () ->
+            return ()
+      )
+
+  let read_file_contents path =
+    let open Lwt_io in
+    with_file ~mode:input ~flags:[O_RDONLY] ~perm:0o000 path @@ fun chan ->
+    read chan
+
+  let save ~contents path =
+    let open Lwt_io in
+    with_file ~mode:output path @@ fun chan -> write chan contents
+
+  let readdir path =
+    path |> Lwt_unix.files_of_directory |> Lwt_stream.to_list >>= fun listing ->
+    List.filter (function "." | ".." -> false | _ -> true) listing
+    |> Lwt.return
+
+  let mkdir_p ?(perm = 0o755) path =
+    let rec loop acc path =
+      file_kind ~follow_symlinks:false path >>= function
+      | Directory ->
+          (* create directories, parents first *)
+          Lwt_list.iter_s (fun f -> f ()) acc
+      | Regular | Other | Unknown ->
+          let msg =
+            Printf.sprintf
+              {|Could not create directory "%s": already exists and it's not a directory|}
+              path
+          in
+          Lwt.fail (Failure msg)
+      | Missing ->
+          let parent = Filename.dirname path in
+          loop ((fun () -> Lwt_unix.mkdir path perm) :: acc) parent
```

Comment:
This could still return EEXIST if something else (whether in this process or elsewhere) races and succeeds creating the directory between checking for existence and creating it.
I think it should be possible to call mkdir directly (without checking first), handle EEXIST by ignoring it, and calling mkdir_p with Filename.parent when it returns ENOENT, and failing on ENOTDIR? (see also how this is implemented in unixext).

Anyway minimal fix here would be to wrap the mkdir with a try/catch and ignore EEXIST.

Reply:
Unixext uses a more pessimistic strategy: it expect to create all parent directories by default, so it does n mkdirs, where n is the number of segments in the path. This strategy is more optimistic in the sense that it expects the path to be there, but doing up to 2n mkdirs. Most of the time the directory will be there in this case, so it makes sense to be optimistic.